### ISSUE TITLE
Provide a stand alone synth driver for eSpeak

### ISSUE THREAD

**Issue by mick-nvaccess** [nvaccess/test_nvdaTracImport#11]
**Reported by jteh on 25 Apr 2007 08:18 UTC**
Currently, [eSpeak](http://espeak.sourceforge.net/) works extremely well with NVDA as a SAPI 5 synthesiser. However, SAPI 5 synthesisers must be registered on the system before they can be used. This means that the user must install eSpeak separately and administrative privileges are required for this installation.

A stand alone eSpeak driver should be provided with NVDA. Not only does this avoid the hastle of installation, but it allows users running NVDA off a portable external drive to use eSpeak on any Windows 2000+ computer without any installation or administrative privileges.

A method needs to be found to interface with eSpeak from NVDA. eSpeak needs to be compiled as a stand alone dll. There are two options:
  * eSpeak uses !PortAudio to provide cross-platform audio output. I'm not sure how well this works under Windows and getting eSpeak to compile against !PortAudio in Windows could be tricky. I suspect this is asynchronous and probably requires POSIX threads (see below).
  * eSpeak could supply audio data to NVDA and NVDA could play the audio data in another thread.
There is also the question of synchronous and asynchronous operation. eSpeak does support asynchronous operation (where the function doesn't block), but this is only supported on platforms with POSIX threads. If we wanted to use asynchronous mode, we would need to adapt eSpeak to use Windows threads.

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 2 May 2007 14:32 UTC**
A driver is nearly complete!
  * We tried a new synchronous playback mode implemented in eSpeak, but it seems that !PortAudio has some issues with lag. Also, attempting to stop speech causes a thread timeout error. I suspect there might be system functions getting executed in an MME callback; apparently, you can't execute such functions in an MME callback without running into deadlock.
  * Instead, I wrote a Python interface to waveOut using ctypes called nvwave. We are now using that with synchronous retrieval mode.
  * The driver is pretty much done, including indexing and waiting, though there are a couple of issues:
    * Upper case pitch change isn't working.
    * Automatic language detection at initialisation might be nice and shouldn't be too hard. We just have to call _espeak.setVoiceByName with the Windows LCID. This should be done instead of the current behaviour of defaulting to English.

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 3 May 2007 10:09 UTC**
First committed in r612. Capital pitch change now works. All is correctly packaged with the binary as of r618. I'm still tempted to implement initial automatic detection of language when switching to the synth for the first time, though. We shouldn't use !SetVoiceByName as I proposed earlier, though, but should use !SetVoiceByProperties with a language code.

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 21 May 2007 05:04 UTC**
Automatic language detection on initialisation implemented by Mick in r723 (r735 in the 0.5 branch).

I can reliably cause an access violation when calling espeak_Initialize() on my system if I initialize, speak and terminate three times. I've yet to work out why this is happening. I suspect it has something to do with memory allocation, but I'd like to fix this if possible before closing this ticket.

**Comment by mick-nvaccess** [comment]
**Comment by jteh on 22 May 2007 03:29 UTC**
Replying to [jteh](comment:3):
> I can reliably cause an access violation when calling espeak_Initialize() on my system if I initialize, speak and terminate three times. I've yet to work out why this is happening.
Fixed in eSpeak 1.25, which is now in trunk.

Currently, there is no way to select voice variants, but I don't think this is necessary for the release of this driver and will require some discussion as to how it should be implemented. This should be handled in a separate ticket.
